File: capa1/simuladores/python/src/main.py
"""
UrbIA — Simulador Python
Publica la telemetría de los medidores del catálogo cuyo `nodo_origen`
coincide con NODE_ID (mono y/o trifásicos) al broker MQTT.
Una instancia por nodo; todo se configura por variables de entorno.
"""
import json
import logging
import os
import signal
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from sensor_monofasico import SensorMonofasico
from sensor_trifasico  import SensorTrifasico
from mqtt_publisher    import MQTTPublisher
from payload_validator import PayloadValidator


def _crear_sensor(m: dict, seed: int):
    cls = SensorTrifasico if m["device_type"] == "trifasico" else SensorMonofasico
    return cls(m["device_id"], seed=seed)

# ── Configuración via variables de entorno ──────────────────────
BROKER_HOST  = os.getenv("BROKER_HOST",   "192.168.0.101")
BROKER_PORT  = int(os.getenv("BROKER_PORT", "1883"))
NODE_ID      = os.getenv("NODE_ID",       "192.168.0.103")
INTERVALO    = float(os.getenv("INTERVALO_SEG", "5"))
LOG_LEVEL    = os.getenv("LOG_LEVEL",     "INFO")
SEED_BASE    = int(os.getenv("SEED_BASE", "42"))

# client_id MQTT: único por instancia. Con clean_session=False dos clientes
# con el mismo id se expulsan mutuamente del broker en bucle, así que el
# default se deriva del último octeto del NODE_ID (.103 → urbia-sim-103) y
# se puede sobreescribir con MQTT_CLIENT_ID.
MQTT_CLIENT_ID = os.getenv(
    "MQTT_CLIENT_ID", f"urbia-sim-{NODE_ID.rsplit('.', 1)[-1]}"
)

logging.basicConfig(
    level=getattr(logging, LOG_LEVEL),
    format="%(asctime)s [%(levelname)s] %(name)s — %(message)s",
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler("/app/logs/simulador.log"),
    ]
)
logger = logging.getLogger("urbia.main")


def cargar_medidores():
    catalogo = json.load(open(
        Path(__file__).parent.parent / "comun" / "medidores_manizales.json"
    ))
    return [m for m in catalogo["medidores"] if m["nodo_origen"] == NODE_ID]


def main():
    medidores_info = cargar_medidores()
    logger.info(f"Nodo {NODE_ID} — {len(medidores_info)} medidores asignados")

    sensores   = [_crear_sensor(m, SEED_BASE + i) for i, m in enumerate(medidores_info)]
    publisher  = MQTTPublisher(BROKER_HOST, BROKER_PORT, MQTT_CLIENT_ID)
    validator  = PayloadValidator()

    publisher.conectar()
    time.sleep(2)
    logger.info(
        f"Conectado a {BROKER_HOST}:{BROKER_PORT} "
        f"(client_id={MQTT_CLIENT_ID}) — iniciando simulación"
    )

    running = [True]
    def stop(sig, frame):
        running[0] = False
        logger.info("Señal de parada recibida")
    signal.signal(signal.SIGTERM, stop)
    signal.signal(signal.SIGINT, stop)

    ciclo = 0
    while running[0]:
        ciclo += 1
        enviados = 0
        for sensor in sensores:
            lectura = sensor.generar_lectura()
            ok, msg = validator.validar(lectura)
            if not ok:
                logger.warning(f"Payload inválido {sensor.device_id}: {msg}")
                continue
            topic = f"urbia/manizales/{lectura['zona']}/{lectura['device_type']}/{lectura['device_id']}/telemetria"
            if publisher.publicar(topic, lectura):
                enviados += 1

        if ciclo % 12 == 0:  # log cada minuto (12 ciclos × 5s)
            stats_pub = publisher.estadisticas()
            stats_val = validator.estadisticas()
            logger.info(
                f"Ciclo {ciclo} — enviados:{enviados}/{len(sensores)} "
                f"total_pub:{stats_pub['publicados']} "
                f"tasa_valid:{stats_val['tasa_exito_pct']}%"
            )

        time.sleep(INTERVALO)

    publisher.desconectar()
    logger.info("Simulador detenido limpiamente")


if __name__ == "__main__":
    main()

File: capa1/simuladores/cpp/src/main.cpp
/**
 * UrbIA — Simulador C++ · Nodo .104
 * Publica telemetría de 50 medidores trifásicos (La Enea y Palermo)
 * al broker MQTT en .101:1883
 */
#include <iostream>
#include <fstream>
#include <vector>
#include <memory>
#include <thread>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <nlohmann/json.hpp>
#include "sensor_trifasico.hpp"
#include "mqtt_publisher.hpp"

static volatile bool running = true;
void signal_handler(int) { running = false; }

int main() {
    signal(SIGTERM, signal_handler);
    signal(SIGINT,  signal_handler);

    const char* broker_host = std::getenv("BROKER_HOST")    ?: "192.168.0.101";
    int broker_port         = std::atoi(std::getenv("BROKER_PORT")    ?: "1883");
    const char* node_id     = std::getenv("NODE_ID")         ?: "192.168.0.104";
    int intervalo           = std::atoi(std::getenv("INTERVALO_SEG")  ?: "5");
    int seed_base           = std::atoi(std::getenv("SEED_BASE")      ?: "42");

    std::ifstream f("/opt/urbia/capa1/comun/medidores_manizales.json");
    if (!f.is_open()) {
        std::cerr << "[ERROR] No se puede abrir medidores_manizales.json\n";
        return 1;
    }
    nlohmann::json catalogo = nlohmann::json::parse(f);

    std::vector<std::unique_ptr<SensorTrifasico>> sensores;
    int idx = 0;
    for (auto& m : catalogo["medidores"]) {
        if (m["nodo_origen"] == node_id) {
            sensores.push_back(std::make_unique<SensorTrifasico>(
                m["device_id"].get<std::string>(), seed_base + idx++
            ));
        }
    }

    if (sensores.empty()) {
        std::cerr << "[ERROR] No hay medidores asignados a nodo " << node_id << "\n";
        return 1;
    }

    std::cout << "[UrbIA C++] Nodo " << node_id
              << " — " << sensores.size() << " sensores cargados\n";

    MQTTPublisher publisher(broker_host, broker_port, "urbia-sim-104");
    try {
        publisher.conectar();
    } catch (const std::exception& e) {
        std::cerr << "[ERROR] " << e.what() << "\n";
        return 1;
    }
    std::this_thread::sleep_for(std::chrono::seconds(2));
    std::cout << "[UrbIA C++] Iniciando simulación — intervalo: "
              << intervalo << "s\n";

    int ciclo = 0;
    while (running) {
        ++ciclo;
        int enviados = 0;

        for (auto& s : sensores) {
            auto lectura = s->generar_lectura();
            std::string zona      = lectura["zona"];
            std::string dev_type  = lectura["device_type"];
            std::string device_id = lectura["device_id"];
            std::string topic = "urbia/manizales/" + zona + "/" +
                                dev_type + "/" + device_id + "/telemetria";
            if (publisher.publicar(topic, lectura))
                ++enviados;
        }

        if (ciclo % 12 == 0) {
            auto stats = publisher.estadisticas();
            std::cout << "[ciclo " << ciclo << "] enviados: "
                      << enviados << "/" << sensores.size()
                      << " | total_pub: " << stats.publicados
                      << " | errores: "   << stats.errores << "\n";
        }

        std::this_thread::sleep_for(std::chrono::seconds(intervalo));
    }

    publisher.desconectar();
    std::cout << "[UrbIA C++] Detenido limpiamente\n";
    return 0;
}
